todo-mvc/src/main/scala/todomvc/TodoMvc.scala

```diff
@@ -154,24 +157,54 @@ object TodoMvc extends IOWebApp:
       )
     )
 
-class TodoStore(entries: SignallingSortedMapRef[IO, Int, Todo], nextId: Ref[IO, Int]):
+class TodoStore(entries: SignallingSortedMapRef[IO, Long, Todo], nextId: IO[Long]):
   def create(text: String): IO[Unit] =
-    nextId.getAndUpdate(_ + 1).flatMap(entries(_).set(Some(Todo(text, false))))
+    nextId.flatMap(entries(_).set(Some(Todo(text, false))))
 
-  def entry(id: Int): SignallingRef[IO, Option[Todo]] = entries(id)
+  def entry(id: Long): SignallingRef[IO, Option[Todo]] = entries(id)
 
-  def ids(filter: Filter): Signal[IO, List[Int]] =
+  def ids(filter: Filter): Signal[IO, List[Long]] =
     entries.map(_.filter((_, t) => filter.pred(t)).keySet.toList)
 
   def size: Signal[IO, Int] = entries.map(_.size)
 
   def activeCount: Signal[IO, Int] = entries.map(_.values.count(!_.completed))
 
 object TodoStore:
-  def empty: IO[TodoStore] =
-    (SignallingSortedMapRef[IO, Int, Todo], IO.ref(0)).mapN(TodoStore(_, _))
 
-case class Todo(text: String, completed: Boolean)
+  def make: Resource[IO, TodoStore] =
+    val key = "todos-calico"
+
+    for
+      mapRef <- SignallingSortedMapRef[IO, Long, Todo].toResource
+
+      _ <- Resource.eval {
+        OptionT(Storage.local[IO].getItem(key))
+          .subflatMap(jawn.decode[SortedMap[Long, Todo]](_).toOption)
+          .foreachF(mapRef.set(_))
+      }
+
+      _ <- Storage
+        .local[IO]
+        .events
+        .foreach {
+          case Storage.Event.Updated(`key`, _, value, _) =>
+            jawn.decode[SortedMap[Long, Todo]](value).foldMapM(mapRef.set(_))
+          case _ => IO.unit
+        }
+        .compile
+        .drain
+        .background
+
+      _ <- mapRef
+        .discrete
+        .foreach(todos => Storage.local[IO].setItem(key, todos.asJson.noSpaces))
+        .compile
+        .drain
+        .backgroundOn(calico.unsafe.MacrotaskExecutor)
+    yield TodoStore(mapRef, IO.realTime.map(_.toMillis))
```

Comment:
On browsers `realTime` is really coarse, like 100-millis coarse. So creating TODOs too rapidly will have identical ids that clobber each other. This will break the benchmark.

The reference TodoMVC seems to be using `Date.now()` as the ids, so it should be reasonable. The difference is that Calico enforces uniqueness of ids.

Edit: huh, maybe I'm wrong about the 100 millis thing. Will do further research ...

Reply:
Ugh I'm utterly stupid 😂 the resolution is 100 _microseconds_, not millis. This is all fine 😂